File: pdoo/ruby/hangar.rb
#encoding:utf-8

require "./lib/HangarToUI.rb"
require_relative "./weapon.rb"
require_relative "./shieldbooster.rb"

module Deepspace
  # Hangar
  class Hangar

    attr_reader :maxElements, :weapons, :shieldBoosters
    
    def initialize(capacity)
      @maxElements = capacity
      @weapons = Array.new()
      @shieldBoosters = Array.new()
    end

    # Constructor por copia
    def self.newCopy(h)
      n_hangar = Hangar.new(h.maxElements)
      h.weapons.each do |weapon|
        n_hangar.addWeapon(weapon)
      end
      h.shieldBoosters.each do |shield|
        n_hangar.addShieldBooster(shield)
      end
      return n_hangar
    end

    def getUIversion
      HangarToUI.new(self)
    end

    def to_s
      getUIversion.to_s
    end

    # Comprueba si hay espacio disponible
    private def spaceAvailable
      if (@weapons.length+@shieldBoosters.length)<@maxElements
        return true
      else
        return false
      end
    end

    # Añade un arma
    def addWeapon(w)
      if spaceAvailable
        @weapons << w
        return true
      end
      return false
    end

    # Añade un potenciador de escudo
    def addShieldBooster(s)
      if spaceAvailable
        @shieldBoosters << s
        return true
      end
      return false
    end

    # Devuelve el arma de índice 'w' y la elimina del hangar
    def removeWeapon(w)
        ret = Weapon.newCopy(@weapons[w])
        @weapons.delete_at(w)
      return ret
     end
    
    # Devuelve el potenciador de escudo de índice 's' y lo elimina del hangar
    def removeShieldBooster(s)
        ret = ShieldBooster.newCopy(@shieldBoosters[s])
        @shieldBoosters.delete_at(s)
      return ret
    end
  end
end

File: pdoo/java/Deepspace/Hangar.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.ArrayList;
        
class Hangar implements Copyable <Hangar> {
    private int maxElements;
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();
    
    Hangar(int capacity) {  maxElements = capacity; }
    
    Hangar(Hangar h) {
        maxElements = h.maxElements;
        weapons = (ArrayList<Weapon>) h.weapons.clone();
        shieldBoosters = (ArrayList<ShieldBooster>) h.shieldBoosters.clone();
    }
    
    HangarToUI getUIversion() { return new HangarToUI(this); }
    
    private boolean spaceAvaiable() {
        return (shieldBoosters.size()+weapons.size())<maxElements;
    }
    
    public boolean addWeapon(Weapon w) {
        if (spaceAvaiable()) {
            weapons.add(w);
            return true;
        } else
            return false;
    }
    
    public boolean addShieldBooster(ShieldBooster s) {
        if (spaceAvaiable()) {
            shieldBoosters.add(s);
            return true;
        } else
            return false;
    }
    
    public int getMaxElements() { return maxElements; }
    
    public ArrayList<ShieldBooster> getShieldBoosters() { return shieldBoosters; }
    
    public ArrayList<Weapon> getWeapons() { return weapons; }
    
    public ShieldBooster removeShieldBooster(int s) {
        if (s<0 || s>=shieldBoosters.size())
            return null;
        else {
            ShieldBooster aux = shieldBoosters.get(s);
            shieldBoosters.remove(s);
            return aux;
        }
    }
    
    public Weapon removeWeapon(int w) {
        if (w<0 || w>=weapons.size())
            return null;
        else {
            Weapon aux = weapons.get(w);
            weapons.remove(w);
            return aux;
        }
    }
    
    @Override
    public String toString() {
        String aux = "Espacios totales: " + maxElements;
        aux += "\nArmas:\n";
        for (Weapon w: weapons)
            aux += w.toString() + "\n";
        aux += "Escudos:\n";
        for (ShieldBooster s: shieldBoosters)
            aux += s.toString() + "\n";
        return aux;
    }

    @Override
    public Hangar copy() {
        return new Hangar(this);
    }
}
